components/ai_chat/core/common/yt_util.cc

```diff
@@ -0,0 +1,246 @@
+    // english auto-generated track, then settle for anything.
+    // TODO(petemill): Consider preferring user's language.
+    auto iter =
+        std::ranges::find_if(caption_tracks, [](const base::Value& track_raw) {
+          const base::Value::Dict* language_track = track_raw.GetIfDict();
+          auto* kind = language_track->FindString("kind");
+          if (kind && *kind == "asr") {
+            return false;
+          }
+          auto* lang = language_track->FindString("languageCode");
+          if (lang && *lang == "en") {
+            return true;
+          }
+          return false;
+        });
+    if (iter == caption_tracks.end()) {
+      iter = std::ranges::find_if(
+          caption_tracks, [](const base::Value& track_raw) {
+            const base::Value::Dict* language_track = track_raw.GetIfDict();
+            auto* lang = language_track->FindString("languageCode");
+            if (lang && *lang == "en") {
+              return true;
+            }
+            return false;
+          });
+    }
+    if (iter == caption_tracks.end()) {
+      iter = caption_tracks.begin();
+    }
+    track = iter->GetIfDict();
+  }
+  if (!track) {
+    return std::nullopt;
+  }
+  const std::string* caption_url_raw = track->FindString("baseUrl");
+
+  if (!caption_url_raw) {
+    return std::nullopt;
+  }
+  return *caption_url_raw;
+}
+
+std::optional<std::string> ParseAndChooseCaptionTrackUrl(
+    std::string_view body) {
+  if (!body.size()) {
+    return std::nullopt;
+  }
+
+  auto result_value =
+      base::JSONReader::ReadAndReturnValueWithError(body, base::JSON_PARSE_RFC);
+
+  if (!result_value.has_value() || result_value->is_string()) {
+    DVLOG(1) << __func__ << ": parsing error: " << result_value.ToString();
+    return std::nullopt;
+  } else if (!result_value->is_dict()) {
+    DVLOG(1) << __func__ << ": parsing error: not a dict";
+    return std::nullopt;
+  }
+
+  auto* caption_tracks = result_value->GetDict().FindListByDottedPath(
+      "captions.playerCaptionsTracklistRenderer.captionTracks");
+  if (!caption_tracks) {
+    DVLOG(1) << __func__ << ": no caption tracks found";
+    return std::nullopt;
+  }
+
+  return ChooseCaptionTrackUrl(*caption_tracks);
+}
+
+std::string ParseYoutubeTranscriptXml(const base::Value& root) {
+  std::string transcript_text;
+
+  // Example Youtube transcript XML:
+  //
+  // <?xml version="1.0" encoding="utf-8"?>
+  // <transcript>
+  //   <text start="0" dur="5.1">Dear Fellow Scholars, this is Two Minute
+  //   Papers with Dr. Károly Zsolnai-Fehér.</text> <text start="5.1"
+  //   dur="5.28">ChatGPT has just been supercharged with  browsing support, I
+  //   tried it myself too,  </text> <text start="10.38" dur="7.38">and I
+  //   think this changes everything. Well, almost  everything, as you will
+  //   see in a moment. And this  </text>
+  // </transcript>
+  if (data_decoder::IsXmlElementNamed(root, "transcript")) {
```

Comment:
Do we still need to support both formats? Do we still get the older version via resource sniffing?

Reply:
I just tested that sniffing by same document navigating and we also got 0 size http response when fetching through the url. So I think sniffing doesn't work anymore and it will instead interfere with the new way because it relies on the content url it passed back and skip the innertube way

Reply:
Yes it should be fine because the key should work across different videos and when we pair that with getting the video ID from the current URL it should also fetch up to date data.

Reply:
removed in https://github.com/brave/brave-core/pull/29619/commits/0248f5990d53a6ce1872e2d8ecc9cea2a2298f3c